packages/data-broker/src/services/DataServiceResolver.js

```diff
@@ -0,0 +1,181 @@
+   * Convenience method
+   * @see getMappingMultipleOrgUnits
+   * @public
+   * @param {DataSource[]} dataSources
+   * @param {string[]} orgUnitCodes
+   * @return {Promise<DataServiceMapping>}
+   */
+  async getMappingByOrgUnitCodes(dataSources, orgUnitCodes) {
+    if (Array.isArray(orgUnitCodes) && orgUnitCodes.length > 0) {
+      const orgUnits = await this.models.entity.find({ code: orgUnitCodes });
+      return this.getMappingMultipleOrgUnits(dataSources, orgUnits);
+    }
+    return this.getMapping(dataSources, undefined);
+  }
+
+  /**
+   * @private
+   * @param {DataElement[]} dataElement
+   * @param {Entity[]} [orgUnits]
+   * @returns {Promise<DataServiceMappingEntry[]>}
+   */
+  async resolveDataElements(dataElements, orgUnits = []) {
+    const countryCodes = orgUnits
+      .map(orgUnit => orgUnit.country_code)
+      .filter(countryCode => countryCode !== null && countryCode !== undefined);
+    const dataElementCodes = dataElements.map(de => de.code);
+
+    const mappings =
+      countryCodes.length > 0
+        ? await this.models.dataElementDataService.find({
+            country_code: countryCodes,
+            data_element_code: dataElementCodes,
+          })
+        : [];
+
+    const resolved = [];
+    for (const dataElement of dataElements) {
+      const deMappings = mappings.filter(m => m.data_element_code === dataElement.code);
+
+      if (deMappings.length > 1) {
+        // If they all resolve to the same data service, this is fine
+        const resolveToTheSameDataService = (mappingA, mappingB) => {
+          if (mappingA.service_type !== mappingB.service_type) return false;
+          return isequal(mappingA.service_config, mappingB.service_config);
+        };
+        for (const mappingA of deMappings) {
+          for (const mappingB of deMappings) {
+            if (mappingA === mappingB) continue;
+            if (!resolveToTheSameDataService(mappingA, mappingB)) {
+              throw new Error(
+                `Conflicting mappings found for Data Source ${
+                  dataElement.code
+                } when fetching for orgUnits ${orgUnits.map(o => o.code).join(',')}`,
+              );
+            }
+          }
+        }
+        // all data element mappings are the same, use first
+        const [deMapping] = deMappings;
+        resolved.push({
+          dataSource: dataElement,
+          service_type: deMapping.service_type,
+          config: deMapping.service_config,
+        });
+        continue;
+      }
+
+      if (deMappings.length === 1) {
+        const [deMapping] = deMappings;
+        resolved.push({
+          dataSource: dataElement,
+          service_type: deMapping.service_type,
+          config: deMapping.service_config,
+        });
+        continue;
+      }
+
+      // No mappings, use default
+      resolved.push({
+        dataSource: dataElement,
+        service_type: dataElement.service_type,
+        config: dataElement.config,
+      });
```

Comment:
```suggestion
      if (deMappings.length === 0) {
        // No mappings, use default
        resolved.push({
          dataSource: dataElement,
          service_type: dataElement.service_type,
          config: dataElement.config,
        });
        continue;
      }
      
        // If they all resolve to the same data service, this is fine
        const resolveToTheSameDataService = (mappingA, mappingB) => {
          if (mappingA.service_type !== mappingB.service_type) return false;
          return isequal(mappingA.service_config, mappingB.service_config);
        };
        for (const mappingA of deMappings) {
          for (const mappingB of deMappings) {
            if (mappingA === mappingB) continue;
            if (!resolveToTheSameDataService(mappingA, mappingB)) {
              throw new Error(
                `Conflicting mappings found for Data Source ${
                  dataElement.code
                } when fetching for orgUnits ${orgUnits.map(o => o.code).join(',')}`,
              );
            }
          }
        }
        
        // all data element mappings are the same, use first
        const [deMapping] = deMappings;
        resolved.push({
          dataSource: dataElement,
          service_type: deMapping.service_type,
          config: deMapping.service_config,
        });
```

Optional, but this cuts down the nesting and line count a bit if you like?

Reply:
Reworked